File: backend-nodejs-ts/src/modules/product/services/ProductService.ts
import { ProductModel, Product } from '../models/ProductModel';

class ProductService {
  create(product: Product) {
    const newProduct = ProductModel.create(product);
    return newProduct;
  }

  public findAll() {
    const products = ProductModel.find({});
    return products;
  }

  findById(id: string) {
    const product = ProductModel.findById(id);
    return product;
  }

  async deleteById(id: string) {
    const product = await this.findById(id);
    if (product) {
      await product.remove();
      return true;
    } else {
      return false;
    }
  }
}

export const productService = new ProductService();

File: backend-springboot/src/main/java/com/backend/spring/modules/product/product/services/ProductService.java
package com.backend.spring.modules.product.product.services;

import com.backend.spring.modules.product.product.entities.Product;
import com.backend.spring.modules.product.review.Review;
import com.backend.spring.modules.user.user.entities.User;

import java.security.Principal;
import java.util.List;
import java.util.Optional;

public interface ProductService {
    List<Product> findAll();

    Optional<Product> findById(Long id);

    Product save(Product product);

    void deleteById(Long id);

    Product createProduct(Principal principal);

    Product updateProduct(Long productId, Product newProduct);

    void addReviewToProduct(Product product, Review review, User user);
}
